app.go

```diff
@@ -35,8 +35,19 @@ import (
 	"go.uber.org/multierr"
 )
 
+type appState int
```

Comment:
i wonder if adding state is extra friction for this PR. Do we want to get just the Invoke part in and then talk about semantics for when other API functions are OK to be called?